plugins/ingestion-kinesis/src/main/java/org/opensearch/plugin/kinesis/KinesisShardConsumer.java

```diff
@@ -0,0 +1,243 @@
+
+import java.io.IOException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.time.Instant;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.TimeoutException;
+
+import static org.opensearch.plugin.kinesis.SequenceNumber.NON_EXISTING_SEQUENCE_NUMBER;
+import static software.amazon.awssdk.auth.credentials.AwsBasicCredentials.create;
+
+/**
+ * Kafka consumer to read messages from a Kafka partition
+ */
+@SuppressWarnings("removal")
+public class KinesisShardConsumer implements IngestionShardConsumer<SequenceNumber, KinesisMessage> {
+    private static final Logger logger = LogManager.getLogger(KinesisShardConsumer.class);
+
+    /**
+     * The Kinesis consumer
+     */
+    ;
+    private KinesisClient kinesisClient;
+    private String lastFetchedSequenceNumber = "";
+    final String clientId;
+    final String kinesisShardId;
+    final int shardId;
+    final KinesisSourceConfig config;
+
+    /**
+     * Constructor
+     * @param clientId the client id
+     * @param config   the kinesis source config
+     * @param shardId the shard id
+     */
+    public KinesisShardConsumer(String clientId, KinesisSourceConfig config, int shardId) {
+        this(clientId, config, shardId, createClient(clientId, config));
+    }
+
+    /**
+     * Constructor, visible for testing
+     * @param clientId the client id
+     * @param config the Kafka source config
+     * @param shardId the shard id
+     * @param kinesisClient the created kinesis client
+     */
+    protected KinesisShardConsumer(String clientId, KinesisSourceConfig config, int shardId, KinesisClient kinesisClient) {
+        this.clientId = clientId;
+        this.kinesisClient = kinesisClient;
+        this.shardId = shardId;
+        this.config = config;
+
+        // Get shard iterator
+        DescribeStreamResponse describeStreamResponse = kinesisClient.describeStream(
+            DescribeStreamRequest.builder().streamName(config.getStream()).build()
+        );
+
+        if (shardId >= describeStreamResponse.streamDescription().shards().size()) {
+            throw new IllegalArgumentException("Shard id " + shardId + " does not exist in stream " + config.getStream());
+        }
+
+        String kinesisShardId = describeStreamResponse.streamDescription().shards().get(shardId).shardId();
+        this.kinesisShardId = kinesisShardId;
+        logger.info("kinesis consumer created for stream {} shard {}", config.getStream(), shardId);
+    }
+
+    /**
+     * Create a Kafka consumer. visible for testing
+     * @param clientId the client id
+     * @param config the Kafka source config
+     * @return the Kafka consumer
+     */
+    protected static KinesisClient createClient(String clientId, KinesisSourceConfig config) {
+
+        KinesisClientBuilder kinesisClientBuilder = KinesisClient.builder()
+            .region(Region.of(config.getRegion()))
+            // TODO: better security config
```

Comment:
Often you'll want to use the `DefaultCredentialsProvider` and possibly include a role that you want to assume. The default provider goes down through a chain of places where credentials could be available, including system properties, environment variables, a profile file (usually $HOME/.aws/credentials), and container/instance-provided.

For testing purposes, environment variables or system properties might work.

Reply:
I took a look at the s3 setup, and the credential management is fairly complicated, and I dont understand it yet. That's why I plan to take it in another PR to reduce the scope. I feel it'll be great if we can make this part a reusable lib for any AWS dependencies, that'll be great

Reply:
Agreed -- the current implementation is fine for now. Someone from AWS can certainly polish it before we're GA.

@andrross mentioned that it would be great to move all AWS plugins into their own repo. That could include a library for common features that all AWS plugins should share, which could include credential management.